### PR TITLE
Cheat function detection

### PR DESCRIPTION
**Pull request by ct222ct**

### CODE REVIEW

File: app/src/main/java/at/aau/edu/wizards/gameModel/GameModelListener.kt [hunk 1]
```diff
@@ -181,8 +181,47 @@ class GameModelListener(
...
     fun updatedGuess(newGuess: Guess) {
-        guesses[activePlayer] = newGuess
+        cheaterPlayerid = newGuess.playerId
+        guesses[newGuess.playerId] = newGuess
+        cheaterPlayerid = newGuess.playerId
+
+        guesses.clear()
+        for (player in players) {
+            for (guess in player.guesses) {
+                guesses.add(Guess(guess, player.id))
```
[the remark this answers is not in the document]
Comment (Kirschzelle) [reply app/src/main/java/at/aau/edu/wizards/gameModel/GameModelListener.kt:193]: (ofc player.hasCheated needs to be set to false again on new round, but you can also just extend the corresponding function in gameModelRules that handles a new round)

**Review by Kirschzelle — CHANGES_REQUESTED** [review]
Alright, first things first. This is a big feature and I'm pretty proud of you. You have come a long way from first working with android.

Okey, so now the big manko, this does not work multiplayer. This is completely local and does not affect the other players at all. Hopefully it is obvious that this is not good, especially since it's a requirement that you implement one feature with networking.

So, I'm going to explain now what you ought to do to get it working on all devices in the game-session. (Under the assumption that you implemented the other feedback ofc.)

First, when a player cheats, it needs to get sent to all other players aswell.

So how do you send stuff to other players?
You use the sendMessage() function, if you call it it will automatically send what you pass it to all players into the receiveMessage() function. (INCLUDING the sending player)
You will need to find a String representation of what you want to send, that is distinct from the other things that already get sent, in a way that you can pick it out after.

Lets try this with your new guessing function. Lets add a new condition legalCheatingGuessSend to the sendMessage() function. If we make a single line string, values 127 and above are still available. Then we could so smth. like this:
(127).toChar().toString() = Player 0 New Guess value 0
(128).toChar().toString() = Player 0 New Guess value 1
...
(137).toChar().toString() = Player 0 New Guess value 11
(138).toChar().toString() = Player 1 New Guess value 0
(139).toChar().toString() = Player 1 New Guess value 1
...
(193).toChar().toString() = Player 6 New Guess value = 11

now we have a representation that we can easily catch out again. You can easily convert the guess to the String value if you play around a bit with * and + math operations.

Now every player gets sent our string representation, lets catch it in our receiveMessage() function. We can simply catch new guesses out with our legalCheatingGuessSend function. Then with String[0].code() we can get the the integer value again. If we play around a bit with % and / Math operators we can easily get our player and new guess value and call our gamemodelrules function that takes care of updateing it. Ta-Daaa now every player has the updated guess and every player model has set his hasCheated to true.

Now just do something similar for when someone calls out a cheater, send the command to every player and then do the action on every player locally.

If you have questions hit me up, but I really tried to explain as detailed as possible.

File: app/src/main/java/at/aau/edu/wizards/gameModel/GameModelListener.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (Kirschzelle) [reply app/src/main/java/at/aau/edu/wizards/gameModel/GameModelListener.kt:193]: jah mb - she just changed the listener value here - not wrong, but as explained above doesn't actually do shit

File: app/src/main/java/at/aau/edu/wizards/gameModel/GameModelListener.kt [hunk 2]
```diff
@@ -181,8 +181,47 @@ class GameModelListener(
         }
     }
 
+    private var cheaterPlayerid: Int = 7
```
[the remark this answers is not in the document]
Comment (Kirschzelle) [reply app/src/main/java/at/aau/edu/wizards/gameModel/GameModelListener.kt:184]: that var should be removed according to my feedback, but yeah, for future - someone who looks at this has no idea why dafuq that is

also small hint - if we got max 6 players and our ids start from 0 - first free value isn't 7

File: app/src/main/java/at/aau/edu/wizards/gameModel/GameModelListener.kt [hunk 3]
```diff
@@ -181,8 +181,47 @@ class GameModelListener(
...
     fun updatedGuess(newGuess: Guess) {
-        guesses[activePlayer] = newGuess
+        cheaterPlayerid = newGuess.playerId
+        guesses[newGuess.playerId] = newGuess
+        cheaterPlayerid = newGuess.playerId
```
[the remark this answers is not in the document]
Comment (ct222ct) [reply app/src/main/java/at/aau/edu/wizards/gameModel/GameModelListener.kt:188]: overlooked during testing and correcting

File: app/src/main/java/at/aau/edu/wizards/ui/scoreboard/ScoreboardFragment.kt [hunk 4]
```diff
@@ -66,6 +65,39 @@ class ScoreboardFragment(val listener: GameModelListener) : Fragment() {
...
+
+
+        binding.btnCheatingFunction.setOnClickListener {
+            var selectedOption = 0
+
+            when {
+                viewModel.listener.guessing -> {
+                    Toast.makeText(activity, "Everyone must have made their guess!", Toast.LENGTH_SHORT).show()
+                }
+                !viewModel.listener.fundCheaterButton -> {
```
[the remark this answers is not in the document]
Comment (ct222ct) [reply app/src/main/java/at/aau/edu/wizards/ui/scoreboard/ScoreboardFragment.kt:77]: Yes, that sounds a lot more sensible than my version

File: app/src/main/java/at/aau/edu/wizards/gameModel/GameModelListener.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (ct222ct) [reply app/src/main/java/at/aau/edu/wizards/gameModel/GameModelListener.kt]: Deleted the variable after I incorporated your feedback

**Pull request merged**